crates/ursa-rpc-service/src/api.rs

```diff
@@ -79,68 +88,228 @@ pub trait NetworkInterface: Sync + Send + 'static {
+            Entry::Occupied(mut e) => {
+                // there is a concurrent request for this cid, just wait for the first one and return
+                e.get_mut().push(tx);
+                return rx
+                    .recv()
+                    .await
+                    .ok_or_else(|| anyhow!("Failed to receive status from channel"))?;
+            }
+            Entry::Vacant(e) => {
+                e.insert(vec![tx]);
             }
         }
-        self.store.blockstore().get(&cid)
+
+        // we are the first concurrent request for this cid
+        let client = self.client.clone();
+        let req = RequestBuilder::new(
+            Method::Get,
+            format!(
+                "https://{}/ipfs/{root_cid}",
+                self.origin_config.ipfs_gateway
+            )
+            .parse()?,
+        )
+        .header("Accept", "application/vnd.ipld.car")
+        .build();
+
+        let store = self.store.db.clone();
+        task::spawn(async move {
+            // send the request
+            let result: Result<(), String> = async {
+                let res = client.send(req).await.map_err(|e| {
+                    format!("Error getting content for cid {root_cid} from origin: {e}")
+                })?;
+
+                load_car_checked(store.as_ref(), res, root_cid)
+                    .await
+                    .map_err(|e| {
+                        format!("Error loading car file for cid {root_cid} from origin: {e}")
+                    })?;
+
+                Ok(())
+            }
+            .await;
+
+            // notify pending requests for cids
+            let mut pending = pending.write().await;
+            match result {
+                Ok(()) => {
+                    if let Some(senders) = pending.remove(&root_cid) {
+                        for sender in senders {
+                            if sender.send(Ok(())).is_err() {
+                                debug!("Failed to send origin status to channel");
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    error!(e);
+                    if let Some(senders) = pending.remove(&root_cid) {
+                        for sender in senders {
+                            if sender.send(Err(anyhow!(e.clone()))).is_err() {
+                                debug!("Failed to send origin status to channel");
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        rx.recv()
+            .await
+            .ok_or_else(|| anyhow!("Failed to receive status from channel"))?
     }
 
-    async fn get_data(&self, root_cid: Cid) -> Result<Vec<(Cid, Vec<u8>)>> {
-        if !self.store.blockstore().has(&root_cid)? {
+    /// Trigger the network and provider to start providing the content id.
+    fn provide_cid(&self, cid: Cid, size: u64) -> JoinHandle<Result<()>> {
```

Comment:
why couldn't we use `put_car` or `put_file`?

Reply:
can't move self

Reply:
we should use `load_car` instead of copying and adding an if statement before the body. Can't async move self? Can you elaborate? I am assuming I can understand correctly. it is because you are spawing a unnecessary task

Reply:
If the delay from provider and network commands are negligible, then I can remove the task and use blocking

Reply:
It doesn't block, we are spawning a task for every action as. Axum for reference https://docs.rs/axum/latest/axum/

Reply:
```
axum -> request1
  |
  -> request2 -> check/fetch to store -> success response
                      \-> put command task
```

I can make this block so that
```
request2 -> check/fetch store -> put command if needed -> response to end user
```

But I think the former is better so that TTFB is lower

Reply:
You have misrepresented how it currently works. It doesn't block each request. I recommended going over the code and Axum again

Reply:
Ok i removed the task https://github.com/fleek-network/ursa/pull/278/commits/fbc8f99c5acc0fce42e47de0f407faea4f53ff37